[CL-1957] Copy content_builder_layouts when copy Project

jinjagit opened the pull request:
Copies `content_builder_layouts` associated with source project, when copying a project either locally (`LocalProjectCopyService`), or via the Project Copy Tool (`ProjectCopyService`).

Also copies any `content_builder_layout_images` (and the related image files) associated with copied layout(s), and ensures their `:code` values match those created in the copied layout(s) to refer to the new images.

Includes tests of the new functionality.

Does not include adding `content_builder_layouts` & `content_builder_layout_images` to tenant template [serializer](https://github.com/CitizenLabDotCo/citizenlab/blob/master/back/engines/commercial/multi_tenancy/app/services/multi_tenancy/templates/serializer.rb), which will be addressed separately in [CL-2746](https://citizenlab.atlassian.net/browse/CL-2746)

Manual testing:
Copies `content_builder_layouts` + (apparently) any number of `layout_images`, with images at any position in layout.
Appears to work with Project copy Tool

[CL-2746]: https://citizenlab.atlassian.net/browse/CL-2746?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jinjagit committed:
[CL-1957] Naive basic copy mechanism - no tests as yet

jinjagit committed:
[CL-1957] Copies all images and inserts new image uuid codes into new layout craftjs json

jinjagit committed:
Merge branch 'master' into CL-1957-Copy-content-builder-layouts-when-copying-project

jinjagit committed:
[CL-1957] Adjust to align with ProjectCopyService in core

jinjagit committed:
[CL-1957] Add tests of copying content builder layouts & images

jinjagit committed:
[CL-1957] Only run tests if Citizenlab.ee?

jinjagit commented on back/spec/services/local_project_copy_service_spec.rb:
```diff
@@ -312,4 +312,243 @@
+        image1 = create(:layout_image)
+        image2 = create(:layout_image)
+
+        craftjs = {
```
Although this large json object could make the test difficult to read, I feel that most people use an IDE where such objects can be collapsed, giving a clearer view:

<img width="918" alt="Screenshot 2023-02-01 at 16 38 22" src="https://user-images.githubusercontent.com/3944042/216106009-f5335f1b-85ef-47d0-aed5-280652c2a3b1.png">

alexander-cit replied:
You can also put this JSON into a file in the `back/spec/fixtures/` folder.

jinjagit replied:
> You can also put this JSON into a file in the `back/spec/fixtures/` folder.

How would I pass in the 2 variables I use in the JSON (2 image code UUIDs)?

alexander-cit replied:
2 ways:
1. Use erb
2. Use this template https://stackoverflow.com/questions/36073829/how-to-use-template-type-strings-in-ruby

jinjagit commented on back/spec/services/local_project_copy_service_spec.rb:
```diff
@@ -312,4 +312,243 @@
+
+            new_image_codes << new_craftjs[locale][node]['props']['dataCode']
+          end
+        end
```
This will probably become a method at the model level (or something like this), that returns a collection of all layout images currently associated with the layout.

This is likely to be useful in many places, including [CL-2753 BE Job to clean up layout images not associated with any layout](https://citizenlab.atlassian.net/browse/CL-2753)

jinjagit committed:
[CL-2746] Use image_url in template for layout_image image

alexander-cit commented on back/app/services/project_copy_service.rb:
```diff
@@ -73,6 +76,54 @@ def export(
+
+      craftjs.each_key do |locale|
+        craftjs[locale].each_key do |node|
+          next unless craftjs[locale][node]['type']['resolvedName'] == 'Image'
```
Could we use this method here? https://github.com/CitizenLabDotCo/citizenlab/blob/cbb5490b52a586f3b414a1c2b06038cb56d58e49/back/engines/commercial/content_builder/app/services/content_builder/layout_service.rb#L11
Even if not, I think it's a good idea to put all the code directly related to the craftjs internal structure in one place, `ContentBuilder::LayoutService`.

alexander-cit commented on back/app/services/project_copy_service.rb:
```diff
@@ -73,6 +76,54 @@ def export(
+          next unless craftjs[locale][node]['type']['resolvedName'] == 'Image'
+
+          source_image_code = craftjs[locale][node]['props']['dataCode']
+          new_image_code = SecureRandom.uuid
```
We could try to keep the logic of generating the code in one place. E.g., in a static method `ContentBuilder::LayoutImage.generate_code`. It can be used here and in the instance method `generate_code` like `self.code = self.class.generate_code`.

jinjagit replied:
Not sure I follow.

We need to generate the UUID here, so we can map it to the old UUID and pass the mapping to the method that creates the yml entries for the images.

If we use a method in the `layout_image` model to generate the UUID, we won't know what the UUID _will_ be when we create the mapping of UUIDs here, which is why I override the code generation in the model, using `unless code`:
```
def generate_code
  self.code = SecureRandom.uuid unless code
end
 ```

alexander-cit replied:
You will also use it here
```ruby
source_image_code = craftjs[locale][node]['props']['dataCode']
new_image_code = ContentBuilder::LayoutImage.generate_code
```

alexander-cit commented on back/app/services/project_copy_service.rb:
```diff
@@ -73,6 +76,54 @@ def export(
+      craftjs = layout.craftjs_jsonmultiloc
+
+      craftjs.each_key do |locale|
+        craftjs[locale].each_key do |node|
```
Little thing, but you could use `each_value do |node|` and use `node` instead of `craftjs[locale][node]`.

jinjagit replied:
Really nice 'little thing'! This cleans up the code nicely. 😄

alexander-cit commented on back/app/services/project_copy_service.rb:
```diff
@@ -73,6 +76,54 @@ def export(
+
+          source_image_code = craftjs[locale][node]['props']['dataCode']
+          new_image_code = SecureRandom.uuid
+          craftjs[locale][node]['props']['dataCode'] = new_image_code
```
That's much better than `gsub` that I proposed initially :+1:

alexander-cit commented on back/spec/services/local_project_copy_service_spec.rb:
```diff
@@ -312,4 +312,243 @@
+
+  if CitizenLab.ee?
+    describe 'when source project has associated content builder layout' do
+      let!(:layout) { create(:layout, code: 'project_description') }
```
Are you sure you need the bang here?

jinjagit replied:
Nice catch! I thought I did, but I don't!

alexander-cit commented on back/spec/services/local_project_copy_service_spec.rb:
```diff
@@ -312,4 +312,243 @@
+        end
+
+        expect(copied_project.content_builder_layouts.first
+          .as_json(except: %i[id content_buildable_id craftjs_jsonmultiloc updated_at created_at]))
```
So, only three columns are compared? Maybe using `only` with them would be clearer. But up to you.

jinjagit replied:
Nice. Didn't know `only` was a thing. I'll use this in multiple other tests here, too.

alexander-cit reviewed (commented):
Good job refactoring our first draft!

jinjagit committed:
[CL-1957] Use exisitng layout image model generate code method to generate UUID

jinjagit committed:
[CL-1957] Use craftjs_element_of_type? to check node type

jinjagit committed:
[CL-1957] Remove unnecessary bang in let statment

jinjagit committed:
[CL-1957] Use each_value to get each craftjs node

jinjagit committed:
[CL-1957] Use only in tests when <= items using except

jinjagit committed:
[CL-1957] Use json.erb fixture in test

jinjagit committed:
[CL-1957] New LayoutService.images method

jinjagit commented on back/app/services/project_copy_service.rb:
```diff
@@ -73,6 +76,54 @@ def export(
+          next unless ContentBuilder::LayoutService.new.craftjs_element_of_type?(node, 'Image')
+
+          source_image_code = node['props']['dataCode']
+          new_image_code = ContentBuilder::LayoutImage.new.generate_code
```
@alexander-cit I _think_ this is what you wanted, but maybe not exactly?

alexander-cit replied:
Almost. This code is good, but this would be even better

```ruby
class ContentBuilder::LayoutImage
  def self.generate_code
    SecureRandom.uuid
  end
  ...
  private
  
  def generate_code
    self.code = self.class.generate_code unless code
  end
 end

...

new_image_code = ContentBuilder::LayoutImage.generate_code
 ```

jinjagit commented:
> Good job refactoring our first draft!

@alexander-cit 

Thanks for all the comments and ideas!

I have addressed all the points you raised and made all the changes you suggested. There is just [one change](https://github.com/CitizenLabDotCo/citizenlab/pull/3850#discussion_r1095723691) I am not sure I made as you imagined.

I also created a new `ContentBuilder::LayoutService#images` method, which returns a collection of all `layout_images` referenced in a `layout` (if they exist), as I can use it here and am pretty sure I will need it when developing a job to clean up dangling `layout_images`

alexander-cit commented on back/spec/services/local_project_copy_service_spec.rb:
```diff
@@ -312,4 +312,50 @@
+
+        image_codes = { code1: image1.code, code2: image2.code }
+        craftjs_str = ERB.new(File.read('spec/fixtures/craftjs_layout_with_2_images.json.erb'))
+          .result(image_codes.instance_eval { binding })
```
Looks good! You can also use `result_with_hash` instead of `result`, it looks a little simpler https://stackoverflow.com/a/44266619

jinjagit replied:
Nice! That helped me clean up that test even more.

jinjagit committed:
[CL-1957] Pass hash to json.erb fixture using result_with_hash

alexander-cit commented on back/engines/commercial/content_builder/app/services/content_builder/layout_service.rb:
```diff
@@ -11,5 +11,19 @@ def select_craftjs_elements_for_type(craftjs, type)
       elt.is_a?(Hash) && elt['type'].is_a?(Hash) && elt.dig('type', 'resolvedName') == type
     end
+
+    def images(layout)
```
I think it's better not to write code for tests. If this method is used only in tests, it can be better to define it right in the test file not to confuse anyone in the future.

jinjagit replied:
> This will probably become a method at the model level (or something like this), that returns a collection of all layout images currently associated with the layout.

> This is likely to be useful in many places, including [CL-2753 BE Job to clean up layout images not associated with any layout](https://citizenlab.atlassian.net/browse/CL-2753)

https://github.com/CitizenLabDotCo/citizenlab/pull/3850#discussion_r1094434808

alexander-cit replied:
> This is likely to be useful in many places, including

Until it's the case, I wouldn't put it in the real code. Because if it's not used, nobody will ever remove it.

jinjagit replied:
OK, removed.

jinjagit committed:
[CL-1957] Improve generate_code method in model

jinjagit committed:
[CL-1957] Remove images method from layout_service
